### PR TITLE
Flatten high-level tunnel forward calls

### PR DESCRIPTION
**Pull request by Wonderful-Space**
Part of https://github.com/ngrok-private/ngrok/issues/21280

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
+## 0.9.0
+* Flattened `tunnel.foward_pipe()` and `tunnel.foward_tcp()` into `tunnel.foward()`. Determination will be made based on `addr` input.
```
Comment (bobzilladev) [inline CHANGELOG.md:2]: nit: s/foward/forward/g

File: examples/ngrok-http-full.py [hunk 2]
```diff
@@ -70,9 +70,11 @@ async def create_tunnel(httpd):
         .metadata("example tunnel metadata from python").listen()
     )
     sock = httpd.server_address
-    tunnel.forward_pipe(sock) if os.name != "nt" else tunnel.forward_tcp(
-        f"localhost:{sock[1]}"
-    )
+    if os.name != "nt":
+        # Set up a unix socket wrapper around standard http server
+        tunnel.foward(sock)
```
Comment (bobzilladev) [inline examples/ngrok-http-full.py:75]: `foward`

File: src/tunnel.rs [hunk 3]
```diff
@@ -413,31 +408,34 @@ make_tunnel_type! {
...
+        info!("Tunnel {id:?} TCP forwarding to {addr:?}");
+        tun.lock().await.fwd_tcp(addr).await
+    };
 
-    debug!("forward_pipe returning");
+    if is_pipe {
+        debug!("forward_pipe returning");
+    } else {
+        debug!("forward_tcp returning");
+    }
```
Comment (bobzilladev) [inline src/tunnel.rs:438]: Feel free to consolidate this down to one debug line, we'll know pipe vs tcp from the info logging above.

File: src/tunnel.rs [hunk 4]
```diff
@@ -413,31 +408,34 @@ make_tunnel_type! {
...
-        .await?
-        .tunnel
-        .lock()
-        .await
-        .fwd_pipe(addr)
-        .await;
+pub async fn forward(id: &String, addr: String) -> PyResult<()> {
+    let tun = &get_storage_by_id(id).await?.tunnel;
+    let is_pipe =
+        addr.starts_with(PIPE_PREFIX) || addr.starts_with("unix:") || addr.starts_with("tun-");
```
Comment (bobzilladev) [inline src/tunnel.rs:414]: The `tun-` is just a convention in the test cases, probably doesn't make sense to bake in here. I do think we should have a `/` to cover fully-qualified unix paths. In the case where a user just has a filename it's hard to differentiate, is `my-project.so` a file or a domain from Somalia?

The most common case are domains, and the user has an out in that they can add `unix:` to force it if we don't auto-detect it happily, so this probably biases in the right direction. Are there other cases we could know it's a path, maybe if there's a forward-slash anywhere in the string? That would cover cases like `./my-project.sock`. They might have weird dns setup so can't rely on "must have a real internet TLD", wouldn't be fun to keep those up to date anyway.

In the end I'd recommend:
1. Automatically assume pipe if the string contains a `/` anywhere (if at the end it won't work as a domain anyway)
2. Remove the notion of `tun-` here. Will have to add `unix:` to the tests
3. Add a note in the comment about `unix:` being a way to force pipe behavior. Maybe that should go in the README too if there's a convenient spot for it.

**Review by bobzilladev — CHANGES_REQUESTED** [review]
looks pretty good to me, a few tweaks in comments

File: CHANGELOG.md (hunk 1 above)
Comment (Wonderful-Space) [reply CHANGELOG.md:2]: Woops 😅

File: src/tunnel.rs [hunk 5]
```diff
@@ -410,17 +412,27 @@ make_tunnel_type! {
...
     let tun = &get_storage_by_id(id).await?.tunnel;
+
+    // You can force a tunnel to be a pipe by prepending "pipe:" to the address.
     let is_pipe =
-        addr.starts_with(PIPE_PREFIX) || addr.starts_with("unix:") || addr.starts_with("tun-");
+        addr.starts_with(PIPE_PREFIX) || addr.starts_with(UNIX_PREFIX) || addr.contains("/");
 
     let res = if is_pipe {
         let mut tun_addr = addr.clone();
+        // Remove any prefix text before "/"
```
Comment (bobzilladev) [inline src/tunnel.rs]: `pipe:` and `unix:` are removed below, what is this meant to handle?

Comment (Wonderful-Space) [reply src/tunnel.rs]: Oh woops, my bad. I thought we were using "/" as a prefix delimiter. 

You meant to just use it as an indicator that it is a pipe.

Comment (bobzilladev) [reply src/tunnel.rs]: Ah, yes! Just using `/` as an indicator. Want `./moo.sock`, `sockets/moo.sock` and `/tmp/moo.sock` to all work and not be modified. Granted we might have to fight more on relative filenames with the upcoming rust changes, but can deal with that when the new rust version is adopted.

**Pull request merged**